File: benchmarks/weird_captcha_gym/shared_runtime/server/incubator_graders/relation_prompt_grounding.py
from __future__ import annotations

import math
from typing import Any


MECHANIC_ID = "relation_prompt_grounding"


def _point(value: Any, width: int, height: int) -> tuple[int, int]:
    if not isinstance(value, list) or len(value) != 2 or any(isinstance(item, bool) for item in value):
        raise ValueError("table point is malformed")
    x, y = int(value[0]), int(value[1])
    if not (0 <= x <= width and 0 <= y <= height):
        raise ValueError("table point leaves assembly stage")
    return x, y


def _inside(point: tuple[int, int], rect: dict[str, Any]) -> bool:
    x, y = point
    return (
        int(rect["x"]) <= x <= int(rect["x"]) + int(rect["width"])
        and int(rect["y"]) <= y <= int(rect["y"]) + int(rect["height"])
    )


def _carousel_point(item: dict[str, Any], tick: int, carousel: dict[str, Any]) -> tuple[int, int]:
    phase = int(item["carousel_phase"])
    ticks = int(carousel["ticks"])
    angle = 2 * math.pi * ((phase + tick) % ticks) / ticks
    center_x, center_y = [int(value) for value in carousel["center"]]
    return (
        round(center_x + int(carousel["radius_x"]) * math.cos(angle)),
        round(center_y + int(carousel["radius_y"]) * math.sin(angle)),
    )


def _snapshot(states: dict[str, dict[str, Any]]) -> list[dict[str, Any]]:
    return [
        {"id": object_id, "x": state["x"], "y": state["y"], "depth": state["depth"], "placed": state["placed"]}
        for object_id, state in sorted(states.items())
    ]


def _browser_round(value: float) -> int:
    """Match JavaScript Math.round, including its negative-half behavior."""
    return math.floor(value + 0.5)


def _relations(
    states: dict[str, dict[str, Any]],
    objects: dict[str, dict[str, Any]],
    constraints: list[dict[str, Any]],
    rules: dict[str, Any],
) -> list[bool]:
    results: list[bool] = []
    for constraint in constraints:
        first = states[str(constraint["a"])]
        second = states[str(constraint["b"])]
        first_object = objects[str(constraint["a"])]
        second_object = objects[str(constraint["b"])]
        kind = str(constraint["type"])
        if kind == "left_of":
            ok = first["x"] + int(rules["horizontal_gap"]) <= second["x"]
        elif kind == "right_of":
            ok = first["x"] >= second["x"] + int(rules["horizontal_gap"])
        elif kind == "inside":
            allowance = int(second_object["radius"]) - int(first_object["radius"]) - int(rules["inside_inset"])
            ok = allowance >= 0 and abs(first["x"] - second["x"]) <= allowance and abs(first["y"] - second["y"]) <= allowance
        elif kind == "in_front_of":
            ok = first["depth"] >= second["depth"] + int(rules["depth_gap"])
        elif kind == "behind":
            ok = first["depth"] + int(rules["depth_gap"]) <= second["depth"]
        else:
            distance = math.hypot(first["x"] - second["x"], first["y"] - second["y"])
            contact_distance = int(first_object["radius"]) + int(second_object["radius"])
            if kind == "touching":
                ok = abs(distance - contact_distance) <= int(rules["touch_tolerance"])
            elif kind == "not_touching":
                ok = distance >= contact_distance + int(rules["not_touch_gap"])
            else:
                ok = False
        results.append(ok)
    return results


def grade(payload: dict[str, Any], ground_truth: dict[str, Any], public_state: dict[str, Any]) -> dict[str, Any]:
    challenge_id = str(ground_truth.get("challenge_id") or "")
    if str(payload.get("mechanic_id") or "") != MECHANIC_ID:
        return {"graded": True, "passed": False, "feedback": "mechanic mismatch"}
    if str(ground_truth.get("mechanic_id") or "") != MECHANIC_ID:
        return {"graded": True, "passed": False, "feedback": "ground-truth mechanic mismatch"}
    if not challenge_id or str(payload.get("challenge_id") or "") != challenge_id:
        return {"graded": True, "passed": False, "feedback": "stale challenge"}
    if str(public_state.get("challenge_id") or "") != challenge_id:
        return {"graded": True, "passed": False, "feedback": "public-state challenge mismatch"}
    task_id = str(ground_truth.get("task_id") or "")
    if not task_id or str(payload.get("task_id") or "") != task_id or str(public_state.get("task_id") or "") != task_id:
        return {"graded": True, "passed": False, "feedback": "task identity mismatch"}
    try:
        stage = ground_truth.get("stage") or {}
        width, height = int(stage["width"]), int(stage["height"])
        carousel = dict(ground_truth.get("carousel") or {})
        worktable = dict(ground_truth.get("worktable_rect") or {})
        objects = {str(item["id"]): dict(item) for item in ground_truth.get("objects") or []}
        if len(objects) != 5:
            raise ValueError("five assembly objects are required")
        projection_targets = {str(item["id"]): dict(item) for item in ground_truth.get("projection_targets") or []}
        if len(projection_targets) != 5 or set(projection_targets) != set(objects):
            raise ValueError("five dual-projection targets are required")
        settle_vectors = {str(key): dict(value) for key, value in (ground_truth.get("settle_vectors") or {}).items()}
        settle_ticks = int(ground_truth.get("settle_ticks"))
        tolerance = {str(key): int(value) for key, value in (ground_truth.get("target_tolerance") or {}).items()}
        if set(tolerance) != {"x", "y", "depth"} or any(value < 0 for value in tolerance.values()):
            raise ValueError("projection tolerance is malformed")
        if public_state.get("projection_targets") != ground_truth.get("projection_targets") or public_state.get("target_tolerance") != ground_truth.get("target_tolerance"):
            raise ValueError("public projection seals differ from hidden contract")
    except (KeyError, TypeError, ValueError) as exc:
        return {"graded": True, "passed": False, "feedback": f"invalid relation contract: {exc}"}

    events = payload.get("events")
    if not isinstance(events, list) or not (1 <= len(events) <= 700):
        return {"graded": True, "passed": False, "feedback": "manipulation transcript is missing or outside limits"}
    states = {
        object_id: {"x": None, "y": None, "depth": int(item["initial_depth"]), "placed": False}
        for object_id, item in objects.items()
    }
    drag: dict[str, Any] | None = None
    depth_drag: dict[str, Any] | None = None
    settle: dict[str, Any] | None = None
    settled = False
    drag_count = 0
    drag_samples = 0
    total_drag_distance = 0.0
    depth_samples = 0
    depth_distance = 0
    settle_samples = 0
    reset_count = 0

    for sequence, event in enumerate(events, start=1):
        if not isinstance(event, dict) or event.get("sequence") != sequence:
            return {"graded": True, "passed": False, "feedback": f"event {sequence} sequence mismatch"}
        kind = str(event.get("kind") or "")
        if kind == "reset":
            states = {
                object_id: {"x": None, "y": None, "depth": int(item["initial_depth"]), "placed": False}
                for object_id, item in objects.items()
            }
            drag = depth_drag = settle = None
            settled = False
            drag_count = drag_samples = depth_samples = depth_distance = settle_samples = 0
            total_drag_distance = 0.0
            reset_count += 1
            continue
        if settled and kind not in {"settle_complete"}:
            return {"graded": True, "passed": False, "feedback": "object manipulation continued after inspection settled"}

        if kind == "drag_start":
            if drag is not None or depth_drag is not None or settle is not None:
                return {"graded": True, "passed": False, "feedback": "overlapping manipulation gestures"}
            object_id = str(event.get("object_id") or "")
            if object_id not in objects:
                return {"graded": True, "passed": False, "feedback": "unknown assembly object"}
            try:
                point = _point(event.get("point"), width, height)
            except ValueError as exc:
                return {"graded": True, "passed": False, "feedback": str(exc)}
            state = states[object_id]
            source = str(event.get("source") or "")
            if not state["placed"]:
                tick = int(event.get("carousel_tick"))
                expected = _carousel_point(objects[object_id], tick, carousel)
                if source != "carousel" or math.hypot(point[0] - expected[0], point[1] - expected[1]) > int(objects[object_id]["radius"]) + 10:
                    return {"graded": True, "passed": False, "feedback": "drag did not begin on the moving carousel object"}
            elif source != "table" or math.hypot(point[0] - state["x"], point[1] - state["y"]) > int(objects[object_id]["radius"]) + 12:
                return {"graded": True, "passed": False, "feedback": "table drag missed the current object geometry"}
            drag = {"object_id": object_id, "start": point, "moves": 0}
            continue
        if kind == "drag_move":
            if drag is None or event.get("object_id") != drag["object_id"]:
                return {"graded": True, "passed": False, "feedback": "drag move has no matching object"}
            try:
                _point(event.get("point"), width, height)
            except ValueError as exc:
                return {"graded": True, "passed": False, "feedback": str(exc)}
            drag["moves"] += 1
            drag_samples += 1
            continue
        if kind == "drag_cancel":
            if drag is None or event.get("object_id") != drag["object_id"]:
                return {"graded": True, "passed": False, "feedback": "drag cancellation has no matching object"}
            drag = None
            continue
        if kind == "drag_end":
            if drag is None or event.get("object_id") != drag["object_id"]:
                return {"graded": True, "passed": False, "feedback": "drag end has no matching object"}
            try:
                point = _point(event.get("point"), width, height)
            except ValueError as exc:
                return {"graded": True, "passed": False, "feedback": str(exc)}
            if drag["moves"] < 2 or not _inside(point, worktable):
                return {"graded": True, "passed": False, "feedback": "object was not physically carried onto the worktable"}
            states[drag["object_id"]].update({"x": point[0], "y": point[1], "placed": True})
            total_drag_distance += math.hypot(point[0] - drag["start"][0], point[1] - drag["start"][1])
            drag_count += 1
            drag = None
            continue
        if kind == "depth_start":
            object_id = str(event.get("object_id") or "")
            if drag is not None or depth_drag is not None or settle is not None or object_id not in states or not states[object_id]["placed"]:
                return {"graded": True, "passed": False, "feedback": "depth rail started without a placed selection"}
            if event.get("value") != states[object_id]["depth"]:
                return {"graded": True, "passed": False, "feedback": "depth rail start value mismatch"}
            depth_drag = {"object_id": object_id, "start": states[object_id]["depth"], "moves": 0}
            continue
        if kind == "depth_move":
            if depth_drag is None or event.get("object_id") != depth_drag["object_id"]:
                return {"graded": True, "passed": False, "feedback": "depth motion has no selected object"}
            value = int(event.get("value"))
            if not 0 <= value <= 100:
                return {"graded": True, "passed": False, "feedback": "depth value outside rail"}
            states[depth_drag["object_id"]]["depth"] = value
            depth_drag["moves"] += 1
            depth_samples += 1
            continue
        if kind == "depth_end":
            if depth_drag is None or event.get("object_id") != depth_drag["object_id"] or event.get("value") != states[depth_drag["object_id"]]["depth"]:
                return {"graded": True, "passed": False, "feedback": "depth rail end mismatch"}
            if depth_drag["moves"] < 2:
                return {"graded": True, "passed": False, "feedback": "depth adjustment was not a physical drag"}
            depth_distance += abs(states[depth_drag["object_id"]]["depth"] - depth_drag["start"])
            depth_drag = None
            continue
        if kind == "settle_start":
            if drag is not None or depth_drag is not None or settle is not None or not all(state["placed"] for state in states.values()):
                return {"graded": True, "passed": False, "feedback": "settle began before all objects were placed"}
            settle = {"next_tick": 1, "elapsed": 0}
            continue
        if kind == "settle_tick":
            if settle is None or event.get("tick") != settle["next_tick"]:
                return {"graded": True, "passed": False, "feedback": "settle tick is out of order"}
            tick = settle["next_tick"]
            for object_id, state in states.items():
                vector = settle_vectors[object_id]
                factor = settle_ticks - tick + 1
                state["x"] += _browser_round(int(vector["dx"]) * factor / settle_ticks)
                state["y"] += _browser_round(int(vector["dy"]) * factor / settle_ticks)
            if event.get("snapshot") != _snapshot(states):
                return {"graded": True, "passed": False, "feedback": f"settle tick {tick} geometry was tampered"}
            elapsed = int(event.get("elapsed_ms"))
            if elapsed < tick * 70 or elapsed < settle["elapsed"]:
                return {"graded": True, "passed": False, "feedback": "settle inspection was not observed for long enough"}
            settle["elapsed"] = elapsed
            settle["next_tick"] += 1
            settle_samples += 1
            continue
        if kind == "settle_complete":
            if settle is None or settle["next_tick"] != settle_ticks + 1:
                return {"graded": True, "passed": False, "feedback": "settle completed before all force ticks"}
            settle = None
            settled = True
            continue
        return {"graded": True, "passed": False, "feedback": f"event {sequence} has unknown kind"}

    final_snapshot = _snapshot(states)
    if payload.get("final_states") != final_snapshot:
        return {"graded": True, "passed": False, "feedback": "submitted assembly state does not match replay"}
    expected = {
        "drag_count": drag_count,
        "drag_samples": drag_samples,
        "depth_samples": depth_samples,
        "depth_distance": depth_distance,
        "settle_samples": settle_samples,
        "reset_count": reset_count,
    }
    for field, value in expected.items():
        if payload.get(field) != value:
            return {"graded": True, "passed": False, "feedback": f"submitted {field} does not match replay"}
    projection_results = {
        object_id: bool(
            state["placed"]
            and abs(int(state["x"]) - int(projection_targets[object_id]["x"])) <= tolerance["x"]
            and abs(int(state["y"]) - int(projection_targets[object_id]["y"])) <= tolerance["y"]
            and abs(int(state["depth"]) - int(projection_targets[object_id]["depth"])) <= tolerance["depth"]
        )
        for object_id, state in states.items()
    }
    passed = (
        settled
        and drag is None
        and depth_drag is None
        and all(state["placed"] for state in states.values())
        and all(projection_results.values())
        and drag_count >= 5
        and drag_samples >= 15
        and total_drag_distance >= 900
        and settle_samples == settle_ticks
    )
    return {
        "graded": True,
        "passed": passed,
        "score": 100 if passed else 0,
        "feedback": (
            f"dual-projection replay: seals {sum(projection_results.values())}/5; placed {sum(state['placed'] for state in states.values())}/5; "
            f"drags {drag_count}; depth travel {depth_distance}; settle ticks {settle_samples}/{settle_ticks}"
        ),
    }


def cheat(public_state: dict[str, Any], ground_truth: dict[str, Any]) -> dict[str, Any]:
    del public_state
    return {
        "projection_targets": ground_truth.get("projection_targets"),
        "solution_positions": ground_truth.get("solution_positions"),
        "settle_vectors": ground_truth.get("settle_vectors"),
    }

File: benchmarks/weird_captcha_gym/shared_runtime/app/mechanics/relation_prompt_grounding.js
(() => {
  "use strict";

  let model = null;

  function clean(value) {
    return String(value == null ? "" : value).replaceAll("&", "&amp;").replaceAll("<", "&lt;").replaceAll(">", "&gt;").replaceAll('"', "&quot;");
  }

  function record(kind, details = {}) {
    const event = {sequence: model.events.length + 1, kind, ...details};
    model.events.push(event);
    return event;
  }

  function stagePoint(event) {
    const rect = document.querySelector(".rel-stage").getBoundingClientRect();
    return [
      Math.max(0, Math.min(model.state.stage.width, Math.round((event.clientX - rect.left) / rect.width * model.state.stage.width))),
      Math.max(0, Math.min(model.state.stage.height, Math.round((event.clientY - rect.top) / rect.height * model.state.stage.height))),
    ];
  }

  function inside(point, rect) {
    return point[0] >= rect.x && point[0] <= rect.x + rect.width && point[1] >= rect.y && point[1] <= rect.y + rect.height;
  }

  function carouselPoint(item) {
    const carousel = model.state.carousel;
    const angle = Math.PI * 2 * ((item.carousel_phase + model.carouselTick) % carousel.ticks) / carousel.ticks;
    return [
      Math.round(carousel.center[0] + carousel.radius_x * Math.cos(angle)),
      Math.round(carousel.center[1] + carousel.radius_y * Math.sin(angle)),
    ];
  }

  function stateSnapshot() {
    return Object.keys(model.objects).sort().map((id) => {
      const item = model.objects[id];
      return {id, x: item.x, y: item.y, depth: item.depth, placed: item.placed};
    });
  }

  function objectPosition(item) {
    return item.placed ? [item.x, item.y] : carouselPoint(item);
  }

  function projectionPoint(item, view) {
    const table = model.state.worktable_rect;
    if (view === "front") return [
      (item.x - table.x) / table.width * 100,
      (item.y - table.y) / table.height * 100,
    ];
    return [item.depth, (item.y - table.y) / table.height * 100];
  }

  function updateProjections() {
    for (const item of Object.values(model.objects)) {
      for (const view of ["front", "side"]) {
        const marker = document.querySelector(`.rel-live-mark[data-object-id="${CSS.escape(item.id)}"][data-view="${view}"]`);
        if (!marker) continue;
        marker.dataset.placed = item.placed ? "true" : "false";
        if (item.placed) {
          const [x, y] = projectionPoint(item, view);
          marker.style.left = `${x}%`;
          marker.style.top = `${y}%`;
        }
      }
    }
  }

  function applyObjectStyles() {
    for (const item of Object.values(model.objects)) {
      const node = document.querySelector(`.rel-object[data-object-id="${CSS.escape(item.id)}"]`);
      if (!node) continue;
      const point = objectPosition(item);
      node.style.left = `${point[0] / model.state.stage.width * 100}%`;
      node.style.top = `${point[1] / model.state.stage.height * 100}%`;
      node.style.setProperty("--rel-size", `${item.radius * 2 / model.state.stage.width * 100}%`);
      node.style.setProperty("--rel-depth-scale", String(.78 + item.depth * .0045));
      node.style.zIndex = String(10 + item.depth);
      node.dataset.placed = item.placed ? "true" : "false";
      node.classList.toggle("is-selected", model.selectedId === item.id);
    }
    const placed = Object.values(model.objects).filter((item) => item.placed).length;
    const placedNode = document.querySelector(".rel-placed-count");
    if (placedNode) placedNode.textContent = `${placed}/5 ON TABLE`;
    updateProjections();
  }

  function updateDepthConsole() {
    const selected = model.objects[model.selectedId];
    const label = document.querySelector(".rel-depth-selection b");
    const value = document.querySelector(".rel-depth-value");
    const knob = document.querySelector(".rel-depth-knob");
    if (label) label.textContent = selected ? selected.label : "NO OBJECT";
    if (value) value.textContent = selected ? String(selected.depth).padStart(3, "0") : "---";
    if (knob) knob.style.top = `${selected ? 100 - selected.depth : 50}%`;
    document.querySelectorAll(".rel-select").forEach((button) => button.classList.toggle("is-selected", button.dataset.objectId === model.selectedId));
  }

  function selectObject(objectId) {
    if (!model.objects[objectId]?.placed || model.settling || model.settled) return;
    model.selectedId = objectId;
    applyObjectStyles();
    updateDepthConsole();
    model.helpers.setReadout(`${model.objects[objectId].label} SELECTED · ADJUST DEPTH OR POSITION`, "idle");
  }

  function dragStart(event) {
    const node = event.target.closest(".rel-object");
    if (!node || model.settling || model.settled || model.submitting || model.terminal) return;
    event.preventDefault();
    document.querySelector(".rel-failure-stamp")?.remove();
    const objectId = node.dataset.objectId;
    const item = model.objects[objectId];
    const point = stagePoint(event);
    record("drag_start", {object_id: objectId, source: item.placed ? "table" : "carousel", carousel_tick: model.carouselTick, point});
    model.drag = {objectId, original: {...item}, start: point, moves: 0};
    model.selectedId = objectId;
    document.querySelector(".rel-stage").setPointerCapture(event.pointerId);
    node.classList.add("is-dragging");
    applyObjectStyles();
    updateDepthConsole();
  }

  function dragMove(event) {
    if (!model?.drag) return;
    event.preventDefault();
    const point = stagePoint(event);
    record("drag_move", {object_id: model.drag.objectId, point});
    model.drag.moves += 1;
    model.dragSamples += 1;
    const item = model.objects[model.drag.objectId];
    item.x = point[0]; item.y = point[1]; item.placed = true;
    applyObjectStyles();
  }

  function dragEnd(event) {
    if (!model?.drag) return;
    event.preventDefault();
    const point = stagePoint(event);
    const drag = model.drag;
    const item = model.objects[drag.objectId];
    document.querySelector(`.rel-object[data-object-id="${CSS.escape(drag.objectId)}"]`)?.classList.remove("is-dragging");
    if (drag.moves >= 2 && inside(point, model.state.worktable_rect)) {
      record("drag_end", {object_id: drag.objectId, point});
      item.x = point[0]; item.y = point[1]; item.placed = true;
      model.dragCount += 1;
      model.helpers.setReadout(`${item.label} PLACED · DEPTH ${item.depth}`, "idle");
    } else {
      record("drag_cancel", {object_id: drag.objectId, point});
      Object.assign(item, drag.original);
      model.helpers.setReadout("DROP CANCELLED · OBJECT RETURNED", "error");
    }
    model.drag = null;
    applyObjectStyles();
    updateDepthConsole();
    updateControls();
  }

  function depthValue(event) {
    const rect = document.querySelector(".rel-depth-track").getBoundingClientRect();
    return Math.max(0, Math.min(100, Math.round(100 - (event.clientY - rect.top) / rect.height * 100)));
  }

  function depthStart(event) {
    const item = model.objects[model.selectedId];
    if (!item?.placed || model.settling || model.settled || model.terminal) return;
    event.preventDefault();
    document.querySelector(".rel-failure-stamp")?.remove();
    record("depth_start", {object_id: item.id, value: item.depth});
    model.depthDrag = {objectId: item.id, start: item.depth, moves: 0};
    document.querySelector(".rel-depth-track").setPointerCapture(event.pointerId);
  }

  function depthMove(event) {
    if (!model?.depthDrag) return;
    event.preventDefault();
    const value = depthValue(event);
    record("depth_move", {object_id: model.depthDrag.objectId, value});
    model.objects[model.depthDrag.objectId].depth = value;
    model.depthDrag.moves += 1;
    model.depthSamples += 1;
    applyObjectStyles();
    updateDepthConsole();
  }

  function depthEnd(event) {
    if (!model?.depthDrag) return;
    event.preventDefault();
    const item = model.objects[model.depthDrag.objectId];
    if (model.depthDrag.moves < 2) {
      const target = depthValue(event);
      for (const value of [Math.round((item.depth + target) / 2), target]) {
        record("depth_move", {object_id: item.id, value});
        item.depth = value;
        model.depthDrag.moves += 1;
        model.depthSamples += 1;
      }
    }
    record("depth_end", {object_id: item.id, value: item.depth});
    model.depthDistance += Math.abs(item.depth - model.depthDrag.start);
    model.depthDrag = null;
    applyObjectStyles(); updateDepthConsole(); updateControls();
    model.helpers.setReadout(`${item.label} DEPTH LOCKED AT ${item.depth}`, "idle");
  }

  function updateControls() {
    const allPlaced = Object.values(model.objects).every((item) => item.placed);
    const settleButton = document.querySelector(".rel-settle");
    if (settleButton) settleButton.disabled = !allPlaced || model.settling || model.settled;
    const status = document.querySelector(".rel-placed-count");
    if (status) status.dataset.ready = allPlaced ? "true" : "false";
  }

  function startSettle() {
    if (model.settling || model.settled || !Object.values(model.objects).every((item) => item.placed)) return;
    document.querySelector(".rel-failure-stamp")?.remove();
    record("settle_start");
    model.settling = true;
    model.selectedId = null;
    updateDepthConsole(); updateControls();
    document.querySelector(".rel-stage")?.classList.add("is-settling");
    model.helpers.setReadout("SETTLE TEST RUNNING · OBSERVE THE FINAL GRAPH", "pending");
    const startedAt = performance.now();
    let tick = 0;
    model.settleTimer = window.setInterval(() => {
      tick += 1;
      for (const item of Object.values(model.objects)) {
        const vector = model.state.settle_vectors[item.id];
        const factor = model.state.settle_ticks - tick + 1;
        item.x += Math.round(vector.dx * factor / model.state.settle_ticks);
        item.y += Math.round(vector.dy * factor / model.state.settle_ticks);
      }
      applyObjectStyles();
      const elapsed = Math.max(Math.round(performance.now() - startedAt), tick * 70);
      record("settle_tick", {tick, elapsed_ms: elapsed, snapshot: stateSnapshot()});
      model.settleSamples += 1;
      const meter = document.querySelector(".rel-settle-meter i");
      if (meter) meter.style.width = `${tick / model.state.settle_ticks * 100}%`;
      if (tick >= model.state.settle_ticks) {
        window.clearInterval(model.settleTimer);
        model.settleTimer = null;
        model.settling = false;
        model.settled = true;
        record("settle_complete", {elapsed_ms: Math.round(performance.now() - startedAt)});
        document.querySelector(".rel-stage")?.classList.remove("is-settling");
        document.querySelector(".rel-stage")?.classList.add("is-settled");
        model.helpers.setReadout("SETTLED · INSPECT THE FINAL RELATION GRAPH", "idle");
        updateControls();
      }
    }, model.state.settle_tick_ms);
  }

  function resetAssembly() {
    if (model.submitting || model.terminal) return;
    document.querySelector(".rel-failure-stamp")?.remove();
    if (model.settleTimer) window.clearInterval(model.settleTimer);
    record("reset");
    model.objects = Object.fromEntries(model.state.objects.map((item) => [item.id, {...item, x: null, y: null, depth: item.initial_depth, placed: false}]));
    model.drag = model.depthDrag = null;
    model.selectedId = null;
    model.settling = model.settled = false;
    model.dragCount = model.dragSamples = model.depthSamples = model.depthDistance = model.settleSamples = 0;
    model.resetCount += 1;
    document.querySelector(".rel-stage")?.classList.remove("is-settling", "is-settled");
    const meter = document.querySelector(".rel-settle-meter i"); if (meter) meter.style.width = "0";
    applyObjectStyles(); updateDepthConsole(); updateControls();
    model.helpers.setReadout("ASSEMBLY RESET · CAROUSEL RESTORED", "idle");
  }

  async function submit() {
    if (model.submitting || model.terminal) return;
    model.submitting = true;
    model.helpers.setReadout("REPLAYING GEOMETRY AND SETTLE FORCES…", "pending");
    try {
      const response = await fetch("/result", {method:"POST",headers:{"content-type":"application/json"},body:JSON.stringify({
        mechanic_id:model.state.mechanic_id,task_id:model.state.task_id,challenge_id:model.state.challenge_id,events:model.events,final_states:stateSnapshot(),
        drag_count:model.dragCount,drag_samples:model.dragSamples,depth_samples:model.depthSamples,depth_distance:model.depthDistance,
        settle_samples:model.settleSamples,reset_count:model.resetCount,completed:model.settled,
      })});
      const outcome = await response.json();
      if (outcome.passed === true) {
        model.terminal = true;
        document.querySelector(".relation-assembly-captcha")?.insertAdjacentHTML("beforeend",'<div class="rel-verdict"><span>DUAL-PROJECTION SCULPTURE CERTIFIED</span><strong>PASS</strong><small>FRONT · SIDE · SETTLE REPLAY VERIFIED</small></div>');
        model.helpers.setReadout("PASS","passed");
      } else if (outcome.passed === false && outcome.state) {
        await model.helpers.render(outcome.state);
        const shell=document.querySelector(".relation-assembly-captcha"); shell?.setAttribute("data-fresh-failure","true"); shell?.insertAdjacentHTML("afterbegin",'<div class="rel-failure-stamp"><b>FAIL</b><span>PROJECTION MISMATCH · FRESH SEALS ISSUED</span></div>');
        const readout=document.querySelector(".readout"); if(readout){readout.textContent="FAIL · FRESH PROJECTION SEALS ISSUED";readout.dataset.status="error";}
      } else { model.submitting=false; model.helpers.setReadout("FAIL · NO AUTHORITATIVE GRADE","error"); }
    } catch (_error) { model.submitting=false; model.helpers.setReadout("FAIL · GEOMETRY VERIFIER OFFLINE","error"); }
  }

  function objectMarkup(item) {
    return `<div class="rel-object rel-shape-${clean(item.shape)} rel-color-${clean(item.color)}" data-object-id="${clean(item.id)}"><div class="rel-object-art"><i></i><b></b></div><span>${clean(item.label)}</span></div>`;
  }

  function projectionMarkup(state, view) {
    const table = state.worktable_rect;
    const point = (item) => view === "front"
      ? [(item.x - table.x) / table.width * 100, (item.y - table.y) / table.height * 100]
      : [item.depth, (item.y - table.y) / table.height * 100];
    return `<div class="rel-projection" data-view="${view}"><header><b>${view === "front" ? "FRONT / X·Y" : "SIDE / Z·Y"}</b><span>TARGET OUTLINE + LIVE SOLID</span></header><div>${state.projection_targets.map((item) => { const [x,y]=point(item); return `<i class="rel-target-mark rel-mini-${clean(item.shape)}" style="left:${x}%;top:${y}%"></i><i class="rel-live-mark rel-mini-${clean(item.shape)}" data-object-id="${clean(item.id)}" data-view="${view}" data-placed="false"></i>`; }).join("")}</div></div>`;
  }

  async function render(state, helpers) {
    if (model?.carouselTimer) window.clearInterval(model.carouselTimer);
    if (model?.settleTimer) window.clearInterval(model.settleTimer);
    document.body.dataset.mechanic="relation-assembly"; document.body.dataset.cheatMode=helpers.isCheatMode()?"true":"false";
    model={state,helpers,objects:Object.fromEntries(state.objects.map((item)=>[item.id,{...item,x:null,y:null,depth:item.initial_depth,placed:false}])),events:[],carouselTick:0,carouselTimer:null,settleTimer:null,drag:null,depthDrag:null,selectedId:null,settling:false,settled:false,dragCount:0,dragSamples:0,depthSamples:0,depthDistance:0,settleSamples:0,resetCount:0,submitting:false,terminal:false};
    window.relationAssemblyModel=model;
    helpers.app.innerHTML=`<section class="relation-assembly-captcha" data-challenge-id="${clean(state.challenge_id)}"><header class="rel-head"><div><span>DUAL-PROJECTION SCULPTURE RIG / ${clean(state.challenge_id)}</span><h1>${clean(state.prompt)}</h1></div><div class="rel-projections">${projectionMarkup(state,"front")}${projectionMarkup(state,"side")}</div></header><main class="rel-main"><section class="rel-stage"><div class="rel-carousel"><span>ROTATING STAGING CAROUSEL</span><i></i></div><div class="rel-table"><span>SCULPTURE WORKTABLE</span><div class="rel-depth-bands"><i>REAR</i><i>MID</i><i>FRONT</i></div></div>${state.objects.map(objectMarkup).join("")}<div class="rel-settle-overlay"><b>FORCE / INSPECTION</b><span>SEALS DESCRIBE THE FINAL SETTLED STATE</span></div></section><aside class="rel-console"><p>DEPTH CALIBRATION</p><h2>Make both camera seals coincide.</h2><div class="rel-depth-selection"><span>SELECTED OBJECT</span><b>NO OBJECT</b></div><div class="rel-depth-rig"><div><span>FRONT / 100</span><div class="rel-depth-track"><i></i><b class="rel-depth-knob"></b></div><span>REAR / 000</span></div><strong class="rel-depth-value">---</strong></div><div class="rel-object-list">${state.objects.map((item)=>`<button type="button" class="rel-select" data-object-id="${clean(item.id)}"><i class="rel-mini-${clean(item.shape)}"></i><span>${clean(item.label)}</span></button>`).join("")}</div><div class="rel-settle-meter"><span>FORCE-SETTLE OBSERVATION</span><b><i></i></b></div><button type="button" class="rel-settle" disabled>RUN FORCE / INSPECTION</button><p class="rel-console-note">Outlined marks are target projections; solid marks are your live sculpture. FRONT fixes X/Y. SIDE fixes depth/Y. Anticipate the visible settle drift.</p></aside></main><footer class="rel-foot"><button type="button" class="rel-reset">RESET SCULPTURE</button><div><span class="rel-placed-count" data-ready="false">0/5 ON TABLE</span><div class="readout" data-status="idle">MOVE EVERY OBJECT FROM THE CAROUSEL</div></div><button type="button" class="rel-submit">${clean(state.submit_label)}</button></footer>${helpers.cheatPanelTemplate()}</section>`;
    const stage=document.querySelector(".rel-stage"); stage.addEventListener("pointerdown",dragStart);stage.addEventListener("pointermove",dragMove);stage.addEventListener("pointerup",dragEnd);stage.addEventListener("pointercancel",dragEnd);
    const depth=document.querySelector(".rel-depth-track");depth.addEventListener("pointerdown",depthStart);depth.addEventListener("pointermove",depthMove);depth.addEventListener("pointerup",depthEnd);depth.addEventListener("pointercancel",depthEnd);
    document.querySelectorAll(".rel-console .rel-select").forEach((button)=>button.addEventListener("click",()=>selectObject(button.dataset.objectId)));
    document.querySelector(".rel-settle").addEventListener("click",startSettle);document.querySelector(".rel-reset").addEventListener("click",resetAssembly);document.querySelector(".rel-submit").addEventListener("click",submit);
    model.carouselTimer=window.setInterval(()=>{model.carouselTick=(model.carouselTick+1)%state.carousel.ticks;applyObjectStyles();},state.carousel.tick_ms);
    applyObjectStyles();updateDepthConsole();updateControls();helpers.installCheatPanel();
  }

  window.WeirdCaptchaMechanics=window.WeirdCaptchaMechanics||{};
  window.WeirdCaptchaMechanics.relation_prompt_grounding={rootSelector:".relation-assembly-captcha",render};
})();
